### PR TITLE
Add Unit Tests for all Parameters() of the service libraries that don't have them.

### PR DESCRIPTION
**Pull request by chandankumar4**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR add the Unit Test cases for all Parameters() of the service libraries that don't have them currently.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2946

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: azure/services/bastionhosts/spec_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			expectedError: "",
+		},
+		{
+			name:     "get BastionHost when all values are present",
+			spec:     &fakeAzureBastionSpec,
+			existing: nil,
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeAssignableToTypeOf(armnetwork.BastionHost{}))
+				g.Expect(result.(armnetwork.BastionHost).Location).To(Equal(ptr.To[string](fakeAzureBastionSpec.Location)))
+				g.Expect(result.(armnetwork.BastionHost).Name).To(Equal(ptr.To[string](fakeAzureBastionSpec.ResourceName())))
```
Comment (sonasingh46) [inline azure/services/bastionhosts/spec_test.go:72]: Do you want to check for `Type` and `SKU` as well?

File: azure/services/natgateways/spec_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (sonasingh46) [inline azure/services/natgateways/spec_test.go:2]: nit: 2023

File: azure/services/roleassignments/spec_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (sonasingh46) [inline azure/services/roleassignments/spec_test.go:2]: nit: 2023

File: azure/services/routetables/spec_test.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (sonasingh46) [inline azure/services/routetables/spec_test.go:2]: nit: 2023

File: azure/services/virtualnetworks/spec_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (sonasingh46) [inline azure/services/virtualnetworks/spec_test.go:2]: nit: 2023

File: azure/services/vnetpeerings/spec_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (sonasingh46) [inline azure/services/vnetpeerings/spec_test.go:2]: nit: 2023

File: azure/services/bastionhosts/spec_test.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (sonasingh46) [inline azure/services/bastionhosts/spec_test.go:2]: nit: 2023

File: azure/services/natgateways/spec_test.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	"context"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4"
+	. "github.com/onsi/gomega"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	natGateway = armnetwork.NatGateway{
```
Comment (sonasingh46) [inline azure/services/natgateways/spec_test.go:29]: You can follow the same naming pattern and instead of `natGateway` you can use `fakeNatGateway`

File: azure/services/natgateways/spec_test.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+			name:     "get NatGateway when all values are present",
+			spec:     &natGatewaySpec1,
+			existing: nil,
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeAssignableToTypeOf(armnetwork.NatGateway{}))
+				g.Expect(result.(armnetwork.NatGateway).Location).To(Equal(ptr.To[string](natGatewaySpec1.Location)))
+				g.Expect(result.(armnetwork.NatGateway).Name).To(Equal(ptr.To[string](natGatewaySpec1.ResourceName())))
+			},
+			expectedError: "",
+		},
```
Comment (sonasingh46) [inline azure/services/natgateways/spec_test.go:75]: What are your thoughts about adding a test case with empty NatGateway i.e  `&armnetwork.NatGateway{}`

File: azure/services/bastionhosts/spec_test.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+			name:     "get BastionHost when all values are present",
+			spec:     &fakeAzureBastionSpec,
+			existing: nil,
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeAssignableToTypeOf(armnetwork.BastionHost{}))
+				g.Expect(result.(armnetwork.BastionHost).Location).To(Equal(ptr.To[string](fakeAzureBastionSpec.Location)))
+				g.Expect(result.(armnetwork.BastionHost).Name).To(Equal(ptr.To[string](fakeAzureBastionSpec.ResourceName())))
+			},
+			expectedError: "",
+		},
```
Comment (sonasingh46) [inline azure/services/bastionhosts/spec_test.go:75]: You can probably add one more test case with empty existing BastionHost

File: azure/services/roleassignments/spec_test.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	roleAssignment = armauthorization.RoleAssignment{
```
Comment (sonasingh46) [inline azure/services/roleassignments/spec_test.go:30]: How about `roleAssignment` --> `fakeRoleAssignment` ?

File: azure/services/roleassignments/spec_test.go [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+		ID:   to.Ptr("fake-id"),
+		Name: to.Ptr("fake-name"),
+		Type: to.Ptr("fake-type"),
+	}
+	fakeRoleAssignment = RoleAssignmentSpec{
```
Comment (sonasingh46) [inline azure/services/roleassignments/spec_test.go:35]: And `fakeRoleAssignment ` --> `fakeRoleAssignmentSpec `

File: azure/services/bastionhosts/spec_test.go (hunk 1 above)
Comment (chandankumar4) [reply azure/services/bastionhosts/spec_test.go:72]: I can add check `SKU` only because `Type` is not used by `Parameters()` method.

File: azure/services/bastionhosts/spec_test.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+			spec:     &AzureBastionSpec{},
+			existing: fakeBastionHost,
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeNil())
+			},
+			expectedError: "",
+		},
+		{
+			name:     "get BastionHost when all values are present",
+			spec:     &fakeAzureBastionSpec,
```
Comment (sonasingh46) [inline azure/services/bastionhosts/spec_test.go:67]: Also, I think it will be more useful if you define this `fakeAzureBastionSpec` here in this file. I know that is defined earlier in `bastionhosts_test.go` file. 

So probably you can do something like : 

```
        fakeSubnetID-1         = "my-subnet-id"
	fakePublicIPID-1       = "my-public-ip-id"
	fakeAzureBastionSpec-1 = AzureBastionSpec{
		Name:        "my-bastion",
		Location:    "westus",
		ClusterName: "my-cluster",
		SubnetID:    fakeSubnetID,
		PublicIPID:  fakePublicIPID,
	}
```

Comment (sonasingh46) [reply azure/services/bastionhosts/spec_test.go:67]: It makes reviewing and navigating the code much easier.

File: azure/services/bastionhosts/spec_test.go (hunk 1 above)
Comment (sonasingh46) [reply azure/services/bastionhosts/spec_test.go:72]: Sure. Then just remove  `Type:     &hostType` from `armnetwork.BastionHost` struct at the top as it is not used anywhere.

File: azure/services/bastionhosts/spec_test.go (hunk 13 above)
Comment (sonasingh46) [reply azure/services/bastionhosts/spec_test.go:67]: Similarly  at other places also in this PR

File: azure/services/natgateways/spec_test.go (hunk 9 above)
Comment (chandankumar4) [reply azure/services/natgateways/spec_test.go:99]: Added

File: azure/services/bastionhosts/spec_test.go (hunk 10 above)
Comment (chandankumar4) [reply azure/services/bastionhosts/spec_test.go:94]: Added the test cases for empty existing BastionHost

File: azure/services/natgateways/spec_test.go (hunk 9 above)
Comment (chandankumar4) [reply azure/services/natgateways/spec_test.go:99]: Here is the one with empty NatGateway (I have to define `Properties` as well otherwise it will fail with nil pointer): https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4021/files#diff-4c03859149181d28bc73b23f73d171b0fbff4be635cbd24d9eb89e016734c138R77

File: azure/services/bastionhosts/spec_test.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+			expect: func(g *WithT, result interface{}) {
+				g.Expect(result).To(BeNil())
+			},
+			expectedError: "",
+		},
+		{
+			name:     "get BastionHost when all values are present",
+			spec:     &fakeAzureBastionSpec1,
+			existing: nil,
+			expect: func(g *WithT, result interface{}) {
```
[the remark this answers is not in the document]
Comment (chandankumar4) [reply azure/services/bastionhosts/spec_test.go:92]: Added the validation for tags

**Review by willie-yao — APPROVED** [review]
/lgtm

Thanks for your work on this! Apologies for the late review.

Note: Can you please squash your commits?
